### PR TITLE
support transformer backbone

### PR DESCRIPTION
**Pull request by xiexinch**
Support transformer backbone.
This PR modified from [https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py](url).

-  Support load pretrained checkpoints from reference repo.
- Training image size can be different from pretrain image.
- Support dynamic test image size at evaluation.

### Usage
Backbone config should be like this.
```python
backbone=dict(
        type='VisionTransformer',
        img_size=(512, 512),
        patch_size=16,
        in_channels=3,
        embed_dim=768,
        depth=12,
        mlp_ratio=4,
        qkv_bias=True,
        qk_scale=None,
        representation_size=None,
        drop_rate=0.0,
        attn_drop_rate=0.0,
        norm_cfg=dict(type='LN'),
        act_cfg=dict(type='GELU'),
        norm_eval=False)
```
We can download checkpoints from  [https://github.com/rwightman/pytorch-image-models](url) by setting pretrained
```python
model = dict(
pretrained='https://github.com/rwightman/pytorch-image-models/releases/download/v0.1-vitjx/jx_vit_base_p16_224-80ecf9dd.pth',
backbone=dict(...))
```

### CODE REVIEW

File: mmseg/models/backbones/vit.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        mlp_hidden_dim = int(dim * mlp_ratio)
+        self.mlp = Mlp(
+            in_features=dim,
+            hidden_features=mlp_hidden_dim,
+            act_cfg=act_cfg,
+            drop=drop)
+
+    def forward(self, x):
+        x = x + self.drop_path(self.attn(self.norm1(x)))
+        x = x + self.drop_path(self.mlp(self.norm2(x)))
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:109]: We may ignore the `drop_path`.

File: mmseg/models/backbones/vit.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        return x
+
+
+class PatchEmbed(nn.Module):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:115]: Missing docstring.

File: mmseg/models/backbones/vit.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+        self.patch_embed = PatchEmbed(
+            img_size=img_size,
+            patch_size=patch_size,
+            in_channels=in_channels,
+            embed_dim=embed_dim)
+        num_patches = self.patch_embed.num_patches
+
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, embed_dim))
+        self.dist_token = nn.Parameter(torch.zeros(
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:205]: These two tokens are not needed.

File: mmseg/models/backbones/vit.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        x = self.proj(x).flatten(2).transpose(1, 2)
+        return x
+
+
+@BACKBONES.register_module()
+class VisionTransformer(nn.Module):
+    """VisionTransformer
+    A PyTorch impl of : `An Image is Worth 16x16 Words: Transformers for
+        Image Recognition at Scale` - https://arxiv.org/abs/2010.11929
+    Includes distillation token & head support for `DeiT: Data-efficient
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:141]: We don't need these two token.

File: mmseg/models/backbones/vit.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
+"""Vision Transformer (ViT) in PyTorch Modified from
+https://github.com/rwightman/pytorch-image-models/blob/master/
+timm/models/vision_transformer.py (Apache-2.0 License)"""
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:3]:
```suggestion
"""Modified from https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py"""
```

File: mmseg/models/backbones/vit.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+from mmcv.runner import load_checkpoint
+from mmcv.utils.parrots_wrapper import _BatchNorm
+
+from mmseg.utils import get_root_logger
+from ..builder import BACKBONES
+
+
+class Mlp(nn.Module):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:20]: Missing docstring.

File: mmseg/models/backbones/vit.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+class Attention(nn.Module):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:45]: Missing docstring.

File: mmseg/models/backbones/vit.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+        x = (attn @ v).transpose(1, 2).reshape(b, n, c)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+        return x
+
+
+class Block(nn.Module):
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:80]: Missing docstring.

File: mmseg/models/backbones/vit.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        embed_dim (int): embedding dimension. Default: 768.
+        depth (int): depth of transformer. Default: 12.
+        num_heads (int): number of attention heads. Default: 12.
+        mlp_ratio (int): ratio of mlp hidden dim to embedding dim. Default: 4.
+        qkv_bias (bool): enable bias for qkv if True. Default: True.
+        qk_scale (float): override default qk scale of head_dim ** -0.5 if set.
+        representation_size (Optional[int]): enable and set representation
+            layer (pre-logits) to this value if set.
+        distilled (bool): model includes a distillation token and head
+            as in DeiT models. Default: False.
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:156]: We don't need to support `distiled`.

File: mmseg/models/backbones/vit.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+        self.norm_eval = norm_eval
+        self.with_cp = with_cp
+        # weight init mode
+        self.weight_init = weight_init
+
+    def init_weights(self, pretrained=None):
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(self, pretrained, strict=False, logger=logger)
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:244]: We may also support https://github.com/rwightman/pytorch-image-models/blob/779107b693010934ac87c8cecbeb65796e218488/timm/models/vision_transformer.py#L417

File: mmseg/models/backbones/vit.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+            # Upsample pos_embed weights if input image size > 224
+            num_pathes = self.img_size // self.patch_size
+            h, w = h // self.patch_size, w // self.patch_size
+            pos_embed = self.pos_embed.reshape(
+                1, num_pathes, num_pathes,
+                self.pos_embed.shape[2]).permute(0, 3, 1, 2)
+            pos_embed = F.interpolate(
+                pos_embed, size=[h, w], align_corners=False, mode='bicubic')
+            self.pos_embed = nn.Parameter(
+                torch.flatten(pos_embed, 2).transpose(1, 2))
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:330]: Updating attributes dynamically during training is not supported by PyTorch.
We may set it during load pretraining.

File: mmseg/models/backbones/vit.py [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+            Default: dict(type='LN', requires_grad=True).
+        act_cfg (dict): Config dict for activation layer.
+            Default: dict(type='GELU').
+        norm_eval (bool): Whether to set norm layers to eval mode, namely,
+            freeze running stats (mean and var). Note: Effect on Batch Norm
+            and its variants only. Default: False.
+        with_cp (bool): (Not Implement) Use checkpoint or not. Using checkpoint
+            will save some memory while slowing down the training speed.
+            Default: False.
+        weight_init: (str): weight init mode.
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:216]: We may add what mode is supported and what does it mean.

File: mmseg/models/backbones/vit.py [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+                OrderedDict([('fc', Linear(embed_dim, representation_size)),
+                             ('act',
+                              build_activation_layer(act_cfg=dict('Tanh')))]))
+        else:
+            self.pre_logits = nn.Identity()
+
+        self.norm_eval = norm_eval
+        self.with_cp = with_cp
+        # weight init mode
+        self.weight_init = weight_init
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:282]: We may add assertation like `assert weight_init in ['jax' , ..]`.

File: mmseg/models/backbones/vit.py [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+                self, pretrained, strict=False, logger=logger)
+            if 'pos_embed' in state_dict.keys(
+            ) and state_dict['pos_embed'].shape != self.pos_embed.shape:
+                self.pos_embed = nn.Parameter(
+                    torch.zeros(state_dict['pos_embed'].shape))
+                logger.info(msg='Reload checkpoint')
+                load_checkpoint(self, pretrained, strict=False, logger=logger)
+                self.pos_embed = nn.Parameter(self.pos_embed[:, 1:, :])
+
+                if self.patch_embed.num_patches != self.pos_embed.shape[1]:
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:297]: We'd better compare `self.pos_embed.shape` against `state_dict['pos_embed'].shape`.

File: mmseg/models/backbones/vit.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+        pos_embed = F.interpolate(
+            pos_embed, size=[h, w], align_corners=False, mode='bicubic')
+        pos_embed = torch.flatten(pos_embed, 2).transpose(1, 2)
+        return pos_embed
+
+    def forward(self, inputs):
+        # print(inputs.shape)
+        x = self.patch_embed(inputs)
+        pos_embed = self.pos_embed
+        if self.img_size != inputs.shape[2:]:
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:351]: We'd better compare shape of the output  and pos_embed.

File: mmseg/models/backbones/vit.py [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+            input_h, input_w = inputs.shape[2:]
+            h, w = self.img_size
+            pos_embed = self.upsample_pos_embed(input_h, input_w, h, w)
+        x = self.pos_drop(x + pos_embed)
+        x = self.blocks(x)
+        x = self.norm(x)
+        x = self.pre_logits(x[:, 0])
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:358]: We don't need pre_logits.

File: mmseg/models/backbones/vit.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+        return tuple([x.reshape(x.shape[0], x.shape[1], 1, 1)])
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:359]: What's the output shape here? It should be [B, C, H, W].

File: mmseg/models/backbones/vit.py [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+                    self.pos_embed = nn.Parameter(pos_embed)
+        elif pretrained is None:
+            normal_init(self.pos_embed)
+            for n, m in self.named_modules():
+                if isinstance(m, Linear):
+                    if n.startswith('pre_logits'):
+                        kaiming_init(m.weight, mode='fan_in')
+                        constant_init(m.bias, 0)
+                    else:
+                        if self.weight_init.startswith('jax'):
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:311]: We may only `jax`.

Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:311]: We may also add references.

File: mmseg/models/backbones/vit.py [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+            logger = get_root_logger()
+            state_dict = load_checkpoint(
+                self, pretrained, strict=False, logger=logger)
+            if 'pos_embed' in state_dict.keys(
+            ) and state_dict['pos_embed'].shape != self.pos_embed.shape:
+                self.pos_embed = nn.Parameter(
+                    torch.zeros(state_dict['pos_embed'].shape))
+                logger.info(msg='Reload checkpoint')
+                load_checkpoint(self, pretrained, strict=False, logger=logger)
+                self.pos_embed = nn.Parameter(self.pos_embed[:, 1:, :])
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:295]: 1. Modify the `"pos_embed"` in `state_dict` inplace.

Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:295]: 2. `load_checkpoint`.

Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:295]: 3. We also add comments.

Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:295]: 4. Log `remove the first pos_embed (corresponding to the cls_token)`.

File: mmseg/models/backbones/vit.py [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+                    kaiming_init(m.weight, mode='fan_in')
+                    if m.bias is not None:
+                        constant_init(m.bias, 0)
+                elif isinstance(m, (_BatchNorm, nn.GroupNorm, nn.LayerNorm)):
+                    constant_init(m.bias, 0)
+                    constant_init(m.weight, 1)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def upsample_pos_embed(self, input_h, input_w, origin_h, origin_w):
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:333]: @staticmethod

File: mmseg/models/backbones/vit.py (hunk 14 above)
Comment (xiexinch) [reply mmseg/models/backbones/vit.py:297]: In this if block, I want to resize the pretrained pos_embed weights to adapt training image, so it should compare `[B, L_training, C]` and` [B, L_pretraining, C]`. In this code, `self.patch_embed.num_patches` refers to `L_trainnig`, and `self.pos_embed.shape[1] `refers to `L_pretraining`.

File: mmseg/models/backbones/vit.py [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+
+    def init_weights(self, pretrained=None):
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            state_dict = load_checkpoint(
+                self, pretrained, strict=False, logger=logger)
+            if 'pos_embed' in state_dict.keys(
+            ) and state_dict['pos_embed'].shape != self.pos_embed.shape:
+                # Resize the pos_embed to pretrained model' pos_embed.
+                # self.pos_embed = nn.Parameter(
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:278]: We may remove the comments.

File: mmseg/models/backbones/vit.py [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+                #     torch.zeros(state_dict['pos_embed'].shape))
+                self.pos_embed = nn.Parameter(state_dict['pos_embed'])
+                logger.info(
+                    msg='Resize the pos_embed to pretrained model pos_embed')
+
+                self.pos_embed = nn.Parameter(self.pos_embed[:, 1:, :])
+                logger.info(msg='Remove the "cls_token" dimension')
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:285]: Could we merge these steps into 1 line?

File: mmseg/models/backbones/vit.py [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+                                 pos_embed.shape)
+            input_h, input_w = img.shape[2:]
+            pos_embed = VisionTransformer.upsample_pos_embed(
+                pos_embed, input_h, input_w, pos_h, pos_w, self.patch_size)
+        return patched_img + pos_embed
+
+    @staticmethod
+    def upsample_pos_embed(pos_embed, input_h, input_w, pos_h, pos_w,
+                           patch_size):
+        """Upsample pos_embed weights."""
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:356]: We may complete this docstring.

File: mmseg/models/backbones/vit.py [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+                elif isinstance(m, (_BatchNorm, nn.GroupNorm, nn.LayerNorm)):
+                    constant_init(m.bias, 0)
+                    constant_init(m.weight, 1)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def _pos_embeding(self, img, patched_img, pos_embed):
+        """Positiong embeding method.
+
+        Resize the pos_embed, it input image size not equal to training
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:320]:
```suggestion
        Resize the pos_embed, if the input image size doesn't match the training size. 
```

File: mmseg/models/backbones/vit.py [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+            img (torch.Tensor): The inference image tensor, the shape
+                must be [B, C, H, W].
+            patched_img (torch.Tensor): The patched image, it should be
+                shape of [B, L1, C].
+            pos_embed (torch.Tensor): The pos_embed weighs, it should be
+                shape of [B, L2, c].
+        Return:
+            torch.Tensor: The pos encoded image feature.
+        """
+        assert len(patched_img.shape) == 3 and len(pos_embed.shape) == 3, \
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:332]:
```suggestion
        assert patched_img.ndim == 3 and pos_embed.ndim == 3, \
```

File: mmseg/models/backbones/vit.py (hunk 12 above)
Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:214]: This argument may not be necessary.

File: mmseg/models/backbones/vit.py [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+        with_cp (bool): (Not Implement) Use checkpoint or not. Using checkpoint
+            will save some memory while slowing down the training speed.
+            Default: False.
+        weight_init: (str): weight init mode.
+    """
+
+    def __init__(self,
+                 img_size=(224, 224),
+                 patch_size=16,
+                 pretrain_img_size=224,
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:220]: We may not need this argument.

File: mmseg/models/backbones/vit.py (hunk 22 above)
Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:285]: We may also log shape is changed from xxx to xxx.

File: mmseg/models/backbones/vit.py (hunk 23 above)
Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:356]: We may use tuple for `(input_h, input_w)` and `(pos_h, pos_w)`.

File: mmseg/models/backbones/vit.py [hunk 27]
```diff
@@ -0,0 +1,386 @@
...
+        if x_len != pos_len:
+            # If run tools/test.py to test a model, the shape of pos_embed
+            # should be [B, L_pretrain, C]
+            if pos_len == (self.pretrain_img_size // self.patch_size)**2:
+                pos_h, pos_w = self.pretrain_img_size, self.pretrain_img_size
+            # During evaluation at training, the shape of pos_embed should be
+            # [B, L_training, C]
+            elif pos_len == (self.img_size[0] // self.patch_size) * (
+                    self.img_size[1] // self.patch_size):
+                pos_h, pos_w = self.img_size
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:344]: These cases could be simplified. 
We need two types of pos_embeding. 
Type 1: Fine-tuning from pretrained backbone. We may infer the new shape of pos embedding from `self.img_size`. 
Type 2: Vanilla forward. No matter training or evaluation, we will resize the pos embedding according to the input image size.

File: mmseg/models/backbones/vit.py (hunk 14 above)
Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:286]: `self.patch_embed.num_patches` is equal to `self.pos_embed.shape[1]`.

File: mmseg/models/backbones/vit.py [hunk 28]
```diff
@@ -0,0 +1,390 @@
...
+        return patched_img + pos_embed
+
+    @staticmethod
+    def resize_pos_embed(pos_embed, input_image_shape, pos_size, patch_size):
+        """Resize pos_embed weights.
+
+        Resize pos_embed using bicubic interpolate method.
+        Args:
+            pos_embed (torch.Tensor): pos_embed weights.
+            input_image_shpae (tuple): Tuple for (input_h, intput_w).
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:355]:
```suggestion
            input_shpae (tuple): Tuple for (input_h, intput_w).
```

File: mmseg/models/backbones/vit.py (hunk 12 above)
Comment (xvjiarui) [reply mmseg/models/backbones/vit.py:215]: ping

**Comment by xvjiarui** [comment]
There are some unresolved comments.

File: mmseg/models/backbones/vit.py [hunk 29]
```diff
@@ -0,0 +1,396 @@
...
+                state_dict = checkpoint['state_dict']
+            else:
+                state_dict = checkpoint
+            # strip prefix of state_dict
+            revise_keys = [(r'^module\.', '')]
+            for p, r in revise_keys:
+                state_dict = {
+                    re.sub(p, r, k): v
+                    for k, v in state_dict.items()
+                }
```
Comment (xvjiarui) [inline mmseg/models/backbones/vit.py:283]: This case could be ignored for now.

**Review by xvjiarui — APPROVED** [review]

**Pull request merged**